nkms_eth/escrow.py

```diff
@@ -1,78 +1,178 @@
 import random
-from typing import List
-from nkms_eth import blockchain
-from nkms_eth import token
-
-ESCROW_NAME = 'MinersEscrow'
-PREMINE = int(1e9) * token.M
-REWARD = token.SATURATION - PREMINE
-HOURS_PER_PERIOD = 1  # 24
-MIN_RELEASE_PERIODS = 1  # 30
-MAX_AWARDED_PERIODS = 365
-MIN_ALLOWED_LOCKED = 10 ** 6
-MAX_ALLOWED_LOCKED = 10 ** 7 * token.M
-MINING_COEFF = [
-    HOURS_PER_PERIOD,
-    2 * 10 ** 7,
-    MAX_AWARDED_PERIODS,
-    MAX_AWARDED_PERIODS,
-    MIN_RELEASE_PERIODS,
-    MIN_ALLOWED_LOCKED,
-    MAX_ALLOWED_LOCKED
-]
-NULL_ADDR = '0x' + '0' * 40
-
-
-def create():
-    """
-    Creates an escrow which manages mining
-    """
-    chain = blockchain.chain()
-    creator = chain.web3.eth.accounts[0]  # TODO: make it possible to override
-    tok = token.get()
-    escrow, tx = chain.provider.get_or_deploy_contract(
-        ESCROW_NAME, deploy_args=[token.get().address] + MINING_COEFF,
-        deploy_transaction={'from': creator})
-    chain.wait.for_receipt(tx, timeout=blockchain.TIMEOUT)
-    tx = tok.transact({'from': creator}).transfer(escrow.address, REWARD)
-    chain.wait.for_receipt(tx, timeout=blockchain.TIMEOUT)
-    tx = escrow.transact({'from': creator}).initialize()
-    chain.wait.for_receipt(tx, timeout=blockchain.TIMEOUT)
-    return escrow
-
-
-def get():
-    """
-    Returns an escrow object
-    """
-    return token.get(ESCROW_NAME)
+from typing import List, Tuple, Set, Generator
+
+from populus.contracts.contract import PopulusContract
+
+from nkms_eth.token import NuCypherKMSToken
+from .blockchain import Blockchain
 
+addr = str
 
-def sample(n: int = 10)-> List[str]:
+
+class Escrow:
     """
-    Select n random staking Ursulas, according to their stake distribution
-    The returned addresses are shuffled, so one can request more than needed and
-    throw away those which do not respond
+    In order to become a participant of the network,
+    a miner locks tokens by depositing to the Escrow contract
+    for a duration measured in periods.
+
+    The contract must be armed, before it can be deployed
+    to the blockchain.
     """
-    escrow = get()
-    n_select = int(n * 1.7)  # Select more ursulas
-    n_tokens = escrow.call().getAllLockedTokens()
-    duration = 10
-
-    for _ in range(5):  # number of tries
-        points = [0] + sorted(random.randrange(n_tokens) for _ in
-                              range(n_select))
-        deltas = [i - j for i, j in zip(points[1:], points[:-1])]
-        addrs = set()
-        addr = NULL_ADDR
-        shift = 0
-
-        for delta in deltas:
-            addr, shift = escrow.call().findCumSum(addr, delta + shift, duration)
-            addrs.add(addr)
-
-        if len(addrs) >= n:
-            addrs = random.sample(addrs, n)
-            return addrs
-
-    raise Exception('Not enough Ursulas')
+
+    _contract_name = 'MinersEscrow'
```

Comment:
...now I see there's a `_contract_name` - so I imagine that this `Escrow` *isn't* literally the contract?

I have the lawyers here on the phone - where do I point them to see "The Contract" that this `Escrow` represents?  And how I do know that?  What is the actual significance of the token `MinersEscrow`?

Reply:
Hah - what is the *real* contract, the bytecode on the blockchain?  You can look under `.deploy()` to see where the the PopulusContract is referenced. "MinersEscrow" is a solidity identifier. http://solidity.readthedocs.io/en/develop/introduction-to-smart-contracts.html#storage

Reply:
So, the way it works in my head is that the bytecode on the blockchain is a valid and executable (ie, legal) version of a Contract, but only worth its weight if it can be confirmed to have been derived (ie, compiled) from a higher-level and more cognizable, reasonable, and readable representation.  Do I have that about right?

If I do, where does `Escrow` fall in those terms?

Reply:
Outside of that process entirely, rather, as a proxy representation and *interface* to the compiled solidity executables.

Reply:
So it's not so much a Contract as it is... an advocate?  a quid pro quo?  a magistrate?  What is this thing really?

Reply:
Hmm, Your point is well taken...